Mathlib/Analysis/Calculus/FDeriv/Mul.lean

```diff
@@ -566,6 +566,112 @@ theorem fderiv_const_mul (ha : DifferentiableAt 𝕜 a x) (b : 𝔸) :
 
 end Mul
 
+section Prod
+
+open BigOperators
+
+/-! ### Derivative of a finite product of functions -/
+
+variable {ι : Type*} [DecidableEq ι] {𝔸' : Type*} [NormedCommRing 𝔸'] [NormedAlgebra 𝕜 𝔸']
+  {u : Finset ι} {f : ι → E → 𝔸'} {f' : ι → E →L[𝕜] 𝔸'}
+
+section Fintype
+
+variable [Fintype ι]
+
+theorem hasStrictFDerivAt_finset_prod_univ {x : ι → 𝔸'} :
+    HasStrictFDerivAt (𝕜 := 𝕜) (∏ i, · i) (∑ i, (∏ j in Finset.univ.erase i, x j) • proj i) x := by
```

Comment:
Should we prove a `List.prod` version for noncommutative normed algebras first, then deduce the `Finset` version from it? BTW, you prove it by generalizing `Finset.univ`, then using induction; why formulate for `univ`, not any `Finset` here?

Reply:
Oh yes, you're right, it makes more sense to just prove for any `Finset`.

I agree, it would be good to have `List.prod` and implement this in terms of that.

I'll revise the PR, thanks!

Reply:
I've pushed a partial first draft. I wasn't sure what form the statements should take for the non-commutative list variants. I've provided these two for `hasStrictFDerivAt_...` (below). The second variant has the advantage that no `Fintype` is required, whereas the first variant supports a more general list.

```lean
theorem hasStrictFDerivAt_list_prod' [Fintype ι] {l : List ι} (hl : l.Nodup) {x : ι → 𝔸} :
    HasStrictFDerivAt (𝕜 := 𝕜) (fun x ↦ (l.map x).prod)
      (∑ i : Fin l.length, ((l.take i).map x).prod •
        smulRight (proj (l.get i)) ((l.drop (.succ i)).map x).prod) x := by ...

theorem hasStrictFDerivAt_list_prod'_finRange {n : ℕ} {x : Fin n → 𝔸} :
    HasStrictFDerivAt (𝕜 := 𝕜) (fun x ↦ ((List.finRange n).map x).prod)
      (∑ i : Fin n, (((List.finRange n).take i).map x).prod •
        smulRight (proj i) (((List.finRange n).drop (.succ i)).map x).prod) x := by ...
```

Then, for the variant that support composition, I have used the statement:

```lean
theorem HasFDerivAt.list_prod' {l : List (E → 𝔸')} {l' : List (E →L[𝕜] 𝔸')} {x : E}
    (hll' : l.length = l'.length)
    (h : ∀ i : Fin l.length,
      HasFDerivAt (fun x ↦ List.get l i x) (List.get l' (i.cast hll')) x) :
    HasFDerivAt (fun x ↦ (l.map fun f ↦ f x).prod)
      (∑ i : Fin l'.length, ((l.take i).map (fun f ↦ f x)).prod •
        smulRight (l'.get i) ((l.drop (.succ i)).map fun f ↦ f x).prod) x := by ...
```

Maybe it would be cleaner to use `{l : List ι} {f : ι → E → 𝔸'} {f' : ι → E →L[𝕜] 𝔸'}` here, or a list of pairs.

Let me know if you have a better idea or an opinion, thanks!

Reply:
Update:

I removed the `HasFDerivAt.func_list_prod'` variant above (with `l : List (E → 𝔸)` a list of functions) and kept only the form with `l : List ι` and `f : ι → E → 𝔸`.

I added the forms `Has{StrictFDeriv,FDeriv,FDerivWithin}At.xxx_prod` and `fderiv[Within]_xxx_prod` for `list_prod'`, `multiset_prod`, `finset_prod`. I didn't provide `list_prod` (for commutative multiplication), as I figured this can just be transformed to into a `Multiset` product using `Multiset.coe_prod`. (There is a single theorem `hasStrictFDerivAt_list_prod` but this is merely a helper to implement `hasStrictFDerivAt_multiset_prod` via `Quotient.inductionOn`.)

While I didn't use it for the other proofs, I added `has[Strict]FDerivAt_list_prod_attach'` that operates on `x : {i // i ∈ l} → 𝔸` in case this is convenient. It might be unnecessary though.

I reduced the proof of `Finset.prod_erase_attach` as suggested, and added the analog `Multiset.erase_attach_map_val` for the `multiset_prod` proof. I also added `List.eraseIdx_eq_take_drop_succ` (perhaps belongs in `Std`?) and `List.erase_get` for the `list_prod'` proof.